logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -0,0 +1,412 @@
+
+    /**
+     * Parsed pipeline configuration graph.
+     */
+    private final PipelineIR graph;
+
+    /**
+     * Ruby pipeline object.
+     */
+    private final RubyIntegration.Pipeline pipeline;
+
+    public CompiledPipeline(final PipelineIR graph, final RubyIntegration.Pipeline pipeline) {
+        this.graph = graph;
+        this.pipeline = pipeline;
+        inputs = setupInputs();
+        filters = setupFilters();
+        outputs = setupOutputs();
+        shutdownFlushes = Collections.unmodifiableList(
+            filters.values().stream().filter(RubyIntegration.Filter::hasFlush)
+                .collect(Collectors.toList())
+        );
+        periodicFlushes = Collections.unmodifiableList(
+            shutdownFlushes.stream().filter(RubyIntegration.Filter::periodicFlush)
+                .collect(Collectors.toList())
+        );
+    }
+
+    public Collection<RubyIntegration.Filter> shutdownFlushers() {
+        return shutdownFlushes;
+    }
+
+    public Collection<RubyIntegration.Filter> periodicFlushers() {
+        return periodicFlushes;
+    }
+
+    public Collection<RubyIntegration.Output> outputs() {
+        return Collections.unmodifiableCollection(outputs.values());
+    }
+
+    public Collection<RubyIntegration.Filter> filters() {
+        return Collections.unmodifiableCollection(filters.values());
+    }
+
+    public Collection<IRubyObject> inputs() {
+        return inputs;
+    }
+
+    public RubyIntegration.Plugin registerPlugin(final RubyIntegration.Plugin plugin) {
+        plugin.register();
+        return plugin;
+    }
+
+    /**
+     * This method contains the actual compilation of the {@link Dataset} representing the
+     * underlying pipeline from the Queue to the outputs.
+     * @return Compiled {@link Dataset} representation of the underlying {@link PipelineIR} topology
+     */
+    public Dataset buildFilterFunc() {
+        return compilePipeline(false);
+    }
+
+    /**
+     * This method contains the actual compilation of the {@link Dataset} representing the
+     * underlying pipeline from the Queue to the outputs as a Debug pipeline.
+     * This means that it will return all {@link JrubyEventExtLibrary.RubyEvent} that passed through
+     * it from its terminal {@link Dataset} by using
+     * {@link Dataset.TerminalDataset.TerminalDebugDataset}.
+     * @return Compiled {@link Dataset} representation of the underlying {@link PipelineIR} topology
+     */
+    public Dataset buildFilterFuncDebug() {
+        return compilePipeline(true);
+    }
+
+    /**
+     * This method contains the actual compilation of the {@link Dataset} representing the
+     * underlying pipeline from the Queue to the outputs.
+     * @param debug True iff called from {@link CompiledPipeline#buildFilterFuncDebug()} to
+     * terminate on a {@link Dataset.TerminalDataset.TerminalDebugDataset}
+     */
+    private Dataset compilePipeline(final boolean debug) {
```

Comment:
(This comment assumes we are keeping the "debug pipeline" construct which seems to exist only for spec suite?)